### ISSUE TITLE
ACLs not honored

### ISSUE THREAD

**Issue by BloodBlight** [lizardfs/lizardfs#816]
I have a LizardFS folder that I am sharing out over SAMBA and having all sorts of issues.  But the big show stopper is that ACLs are not being honored.

When I test on an EXT4 location:
`setfacl -Rm g:USER :rwx /mnt/Test`

And then as that user:
`ls /mnt/Test`
I get:
`-rwxrwx---+ 1 OtherUSER OtherGROUP 9 Feb 15 09:13 /mnt/Test/Test.txt`

However, when I run the same command on a lizardFS share:
`setfacl -Rm g:USER :rwx /LizardFS/Test`
I get:
`ls: cannot access '/LizardFS/Test': Permission denied`

I have the mfsacl option on the mount, but it doesn’t show up when looking at the mount report:
`mfs#mfsmaster:9421 on /LizardFS type fuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other)
`

And if I:
`getfacl /LizardFS/Test/`

I get:
```
# file: LizardFS/Test/
# owner: OtherUSER
# group: OtherGROUP
# flags: -s-
user::rwx
user:OtherUSER:rwx
user:USER:rwx
group::rwx
group:OtherGROUP:rwx
mask::rwx
other::---
default:user::rwx
default:group::rwx
default:mask::rwx
default:other::---
default:other::---
```

My client version is:
`bionic,now 3.12.0+dfsg-1 amd64 [installed]`

Running on Ubuntu 18.04:
`Linux SERVERNAME 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

Ideas?

I am also having an issue with groups not being honored at all, only users.  But that is not a LizardFS issue.

**Comment by oxide94** [comment]
Related: https://github.com/moosefs/moosefs/issues/189

**Comment by BloodBlight** [comment]
Double checked, I already have ignoregid in my exports.

`192.168.0.XXX      /Shares/Test         rw,alldirs,maproot=0,ignoregid`

Not sure if I am missing something.

The master server is not domain joined, only the client.

**Comment by pbeza** [comment]
> I have the mfsacl option on the mount, but it doesn’t show up when looking at the mount report:
> `mfs#mfsmaster:9421 on /LizardFS type fuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other) `

Note that `mfsmount --help` states that `mfsacl` mount option [is][1] deprecated:

    -o mfsacl                   DEPRECATED, used to enable/disable ACL support, ignored now

that's why it doesn't show up in te mount report. ACL is always enabled since commit [`d6511fd77e9f4397f20c9f2bb15f56b03d370405`][2].

    commit d6511fd77e9f4397f20c9f2bb15f56b03d370405
    Author: Piotr Sarna <sarna@skytechnology.pl>
    Date:   Fri Sep 8 11:31:56 2017 +0200
    
        all: Deprecate enable_acl flag
        
        This commit makes mount and client lib ignore enable_acl/mfsacl flag,
        as ACL support is always enabled. Setting the -o mfsacl parameter
        in mount is still possible in order to provide backwards compatibility.
        
        Change-Id: I550d34e2e87d83c6ed895523766d98c56c59861a

I'm not able to reproduce your problem:

![Peek 2019-04-30 11-52](https://user-images.githubusercontent.com/3036218/56954186-7d8c4580-6b3e-11e9-899d-e294cfc8daf7.gif)

You can try to remove `ignoregid` mount option. You can also run `mfsmount` with `--nostdopts`. This will [not][3] add standard LizardFS mount options. Please write back if it helped.

[1]: https://github.com/lizardfs/lizardfs/blob/master/src/mount/fuse/mount_config.h#L138
[2]: https://github.com/lizardfs/lizardfs/commit/d6511fd77e9f4397f20c9f2bb15f56b03d370405
[3]: https://github.com/lizardfs/lizardfs/blob/master/src/mount/fuse/mount_config.h#L42

**Comment by BloodBlight** [comment]
It was not that the ACLs where not applied and shown, it was that they where not honored.  So granting a user permission would not actually work even though the file showed that the user had access.

1. Create an unprivileged user and a group for that user.
2. Create a folder and set the owner to root:root.
3. Create a file with some text in the folder and confirm it is also root:root.
4. Set the perms on that folder and file to something like 770 (or what ever you prefer, just so that "other" has no permissions).
5. Then add that users group (not the user) to the folder by ACL.
6. Test accessing the folder as that user.

**Comment by pbeza** [comment]
OK, now I get it. I confirm the problem.

Test live preview:

![Peek 2019-05-02 09-06](https://user-images.githubusercontent.com/3036218/57060499-f06dfb80-6cb9-11e9-93ab-d351121b89e2.gif)

Test log:

    root@lizardfs-client:~# ls -la /mnt
    total 16
    drwxr-xr-x 1 root root 4096 Apr 26 08:27 .
    drwxr-xr-x 1 root root 4096 Apr 26 07:22 ..
    drwxr-xr-x 2 root root 4096 Apr 26 07:22 lizardfs
    drwxr-xr-x 2 root root 4096 Apr 26 08:27 lizardfs-test
    root@lizardfs-client:~# mount | grep -i lizardfs
    root@lizardfs-client:~# cat /etc/fstab
    lizardfs-master: /mnt/lizardfs lzfs defaults 0 0
    root@lizardfs-client:~# mount /mnt/lizardfs
    root@lizardfs-client:~# mount | grep -i lizard
    lizardfs-master:9421 on /mnt/lizardfs type fuse.mfs (rw,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other)
    root@lizardfs-client:~# ls -la /mnt
    total 12
    drwxr-xr-x  1 root root 4096 Apr 26 08:27 .
    drwxr-xr-x  1 root root 4096 Apr 26 07:22 ..
    drwxrwxrwx+ 3 root root    0 May  2 07:04 lizardfs
    drwxr-xr-x  2 root root 4096 Apr 26 08:27 lizardfs-test
    root@lizardfs-client:~# mkdir /mnt/lizardfs/test
    root@lizardfs-client:~# for i in {1,2,3}; do echo "Test 0${i}" > /mnt/lizardfs/test/test_0${i}.txt ; done
    root@lizardfs-client:~# chmod -R 770 /mnt/lizardfs/test
    root@lizardfs-client:~# setfacl -m g:test:rwx /mnt/lizardfs/test/
    root@lizardfs-client:~# ls -la /mnt
    total 12
    drwxr-xr-x  1 root root 4096 Apr 26 08:27 .
    drwxr-xr-x  1 root root 4096 Apr 26 07:22 ..
    drwxrwxrwx+ 4 root root    0 May  2 07:05 lizardfs
    drwxr-xr-x  2 root root 4096 Apr 26 08:27 lizardfs-test
    root@lizardfs-client:~# ls -la /mnt/lizardfs
    total 4
    drwxrwxrwx+ 4 root root    0 May  2 07:05 .
    drwxr-xr-x  1 root root 4096 Apr 26 08:27 ..
    drwxr-xr-x  8 root root    0 May  2 06:20 duplicates
    drwxrwx---+ 2 root root    0 May  2 07:05 test
    root@lizardfs-client:~# ls -la /mnt/lizardfs/test
    total 2
    drwxrwx---+ 2 root root 0 May  2 07:05 .
    drwxrwxrwx+ 4 root root 0 May  2 07:05 ..
    -rwxrwx---  1 root root 8 May  2 07:05 test_01.txt
    -rwxrwx---  1 root root 8 May  2 07:05 test_02.txt
    -rwxrwx---  1 root root 8 May  2 07:05 test_03.txt
    root@lizardfs-client:~# getfacl -p /mnt/
    # file: /mnt/
    # owner: root
    # group: root
    user::rwx
    group::r-x
    other::r-x
    
    root@lizardfs-client:~# getfacl -p /mnt/lizardfs
    # file: /mnt/lizardfs
    # owner: root
    # group: root
    user::rwx
    group::rwx
    other::rwx
    
    root@lizardfs-client:~# getfacl -p /mnt/lizardfs/test/
    # file: /mnt/lizardfs/test/
    # owner: root
    # group: root
    user::rwx
    group::rwx
    group:test:rwx
    mask::rwx
    other::---
    
    root@lizardfs-client:~# su test
    test@lizardfs-client:/root$ ls -la /mnt/lizardfs
    total 4
    drwxrwxrwx+ 4 root root    0 May  2 07:05 .
    drwxr-xr-x  1 root root 4096 Apr 26 08:27 ..
    drwxr-xr-x  8 root root    0 May  2 06:20 duplicates
    drwxrwx---+ 2 root root    0 May  2 07:05 test
    test@lizardfs-client:/root$ ls -la /mnt/lizardfs/test
    ls: cannot open directory '/mnt/lizardfs/test': Permission denied
    test@lizardfs-client:/root$ getfacl -p /mnt/lizardfs
    # file: /mnt/lizardfs
    # owner: root
    # group: root
    user::rwx
    group::rwx
    other::rwx
    
    test@lizardfs-client:/root$ getfacl -p /mnt/lizardfs/test
    # file: /mnt/lizardfs/test
    # owner: root
    # group: root
    user::rwx
    group::rwx
    group:test:rwx
    mask::rwx
    other::---

**Comment by pbeza** [comment]
[Here is][1] bugfix for `mfsmount3`.

[1]: http://cr.skytechnology.pl:8081/#/c/3337/1/src/mount/fuse/main.cc

**Comment by pbeza** [comment]
For now problem was solved for `mfsmount3` only (see: [proof][1]). I'm not sure whether it is still possible to support ACL in FUSE 2.9.9 (which is provided by [`libfuse2`][2] package) after [changes][3] in kernel `4.9`. For details refer what Miklos Szeredi answered in [this][4] `fuse-devel` mailing list thread.

If you have any suggestions how this issue can be fixed using FUSE 2.9.9, please share your thoughts. ;-)

[1]: https://user-images.githubusercontent.com/3036218/58029319-9758f100-7b1c-11e9-8de3-6517e9dbab0f.gif
[2]: https://packages.ubuntu.com/disco/libfuse2
[3]: https://github.com/torvalds/linux/commit/60bcc88ad185d512f5718f2f8dcccb483ea8fb73
[4]: https://sourceforge.net/p/fuse/mailman/message/36678016/

**Comment by trapexit** [comment]
It's not possible to enable newer features in 2.9.x as there is no access to the INIT instruction's outarg flags variable. There's no way to just define the flag yourself and include it.

https://github.com/libfuse/libfuse/blob/fuse_2_9_bugfix/lib/fuse_lowlevel.c#L1846

It might be possible if you rewrote the whole low level init function and tried to short circuit it but I've not looked close enough at the code to tell if that's plausible.

**Comment by pbeza** [comment]
Thanks @trapexit for your comment – it was important for me to make sure that it's not possible to support ACL using `libfuse 2.9.x` without modifying Linux kernel.

BTW: I've just found 2 patches that modify both kernel and `libfuse 2.9.x` to support ACLs. Here they are:
- https://github.com/lizardfs/lizardfs/blob/master/utils/libfuse_acl.patch
- https://github.com/lizardfs/lizardfs/blob/master/utils/linux_fuse_acl.patch

To sum up: if anyone want to use `libfuse 2.9.x` with ACLs, then sorry – you need to patch your kernel and `libfuse` or upgrade to `libfuse3`.

**Comment by trapexit** [comment]
Not sure I follow. The kernel isn't the problem when talking about libfuse2. The problem is that libfuse2 itself provides no way to ask the kernel to enable that feature as it provides hard coded set of supported FUSE features and gives no raw access to the value returned back to the kernel during the handshake.

**Comment by BloodBlight** [comment]
How is MooseFS handling this? It doesn't seem to be affected.

**Comment by trapexit** [comment]
I'm not familiar with MooseFS or LizardFS but it appears to me they are managing posix acls entirely manually. https://github.com/moosefs/moosefs/blob/6bd69b6fa45806c49f130212c914150e6c9c17d6/mfsclient/mfs_fuse.c#L5277

**Comment by pbeza** [comment]
> Not sure I follow. The kernel isn't the problem when talking about libfuse2. The problem is that libfuse2 itself provides no way to ask the kernel to enable that feature as it provides hard coded set of supported FUSE features and gives no raw access to the value returned back to the kernel during the handshake.

ACL support for FUSE3 [was added][1] in kernel 4.9 – that's why I mentioned that you need to upgrade the kernel to be able to use `libfuse3` with ACLs.

There are 2 scenarios – if you have `libfuse 2.9.x` and:
- Kernel older than `4.9`, you need those [two][2] [patches][3] to patch both the kernel and `libfuse 2.9.x`. Those patches add [`.get_acl`][4] handler to the kernel's [`inode_operations`][5] structure (this patch is no longer needed for kernels newer than `4.9` because [`.get_acl`][6] was added to the `kernel 4.9`) and [provide][8] a way to enable ACL in userspace while mounting a filesystem.
- Kernel at least in version `4.9`, you need `libfuse3` which provides a way to ask the kernel to use ACLs by setting [`FUSE_CAP_POSIX_ACL`][7] flag.

> I'm not familiar with MooseFS or LizardFS but it appears to me they are managing posix acls entirely manually. https://github.com/moosefs/moosefs/blob/6bd69b6fa45806c49f130212c914150e6c9c17d6/mfsclient/mfs_fuse.c#L5277

That is actually the answer for my question if it is possible to support ACL using `libfuse 2.9.x` without modifying Linux kernel. So the answer is _yes, there is such an option_ however IMO leaving ACL handling to the kernel is more reasonable than handling it manually.

[1]: https://github.com/torvalds/linux/commit/60bcc88ad185d512f5718f2f8dcccb483ea8fb73
[2]: https://github.com/lizardfs/lizardfs/blob/master/utils/libfuse_acl.patch
[3]: https://github.com/lizardfs/lizardfs/blob/master/utils/linux_fuse_acl.patch
[4]: https://github.com/lizardfs/lizardfs/blob/master/utils/linux_fuse_acl.patch#L76
[5]: https://elixir.bootlin.com/linux/latest/source/include/linux/fs.h#L1831
[6]: https://github.com/torvalds/linux/commit/60bcc88ad185d512f5718f2f8dcccb483ea8fb73#diff-17976afbf65fce58e63212466ba358f8R1800
[7]: https://github.com/libfuse/libfuse/blob/master/include/fuse_common.h#L327
[8]: https://github.com/lizardfs/lizardfs/blob/master/utils/libfuse_acl.patch#L11

**Comment by trapexit** [comment]
This is not the most popular course of action but with my project I ended up vendoring / embedding libfuse into my project because some of my users had old version of libfuse2 which were buggy, updating libfuse2 on said platforms was non-trivial for my users, libfuse3 wasn't available popularly at the time, and even when libfuse3 was available it didn't always have updates to support the newest features (at least the version included in the distro.)
